linalg/batchlinalg.cpp

```diff
@@ -0,0 +1,391 @@
+                                                    lu_ptr_array_.Read(),
+                                                    matrix_size_,
+                                                    vector_array.ReadWrite(),
+                                                    matrix_size_,
+                                                    num_matrices_);
+
+      MFEM_VERIFY(status_upp == MFEM_CU_or_HIP(BLAS_STATUS_SUCCESS),
+                  "Failed at blasDtrsmBatched upper");
+   }
+   else
+#endif
+   {
+      BatchLUSolve(LUMatrixBatch_, P_, x);
+   }
+
+}
+
+//hand rolled block mult
+void ApplyBlkMult(const mfem::DenseTensor &Mat, const mfem::Vector &x,
+                  mfem::Vector &y)
+{
+   const int ndof = Mat.SizeI();
+   const int NE   = Mat.SizeK();
+   auto X         = mfem::Reshape(x.Read(), ndof, NE);
+   auto Y         = mfem::Reshape(y.Write(), ndof, NE);
+   auto Me        = mfem::Reshape(Mat.Read(), ndof, ndof, NE);
+
+   //Takes row major format
+   mfem::forall(ndof* NE, [=] MFEM_HOST_DEVICE (int tid)
+   {
+
+      const int c = tid % ndof;
+      const int e = tid / ndof;
+
+      {
+         double dot = 0;
+         for (int r = 0; r < ndof; ++r)
+         {
+            dot += Me(r, c, e) * X(r, e);
+         }
+         Y(c, e) = dot;
+      }
+   });
+}
+
+
+void LibBatchSolver::ApplyInverse(const mfem::Vector &b, mfem::Vector &x) const
+{
+   //Extend with vendor library capabilities
+   ApplyBlkMult(InvMatrixBatch_, b, x);
+}
+
+
+void LibBatchSolver::Setup()
+{
+   matrix_size_  = LUMatrixBatch_.SizeI();
+   num_matrices_ = LUMatrixBatch_.SizeK();
+
+   P_.SetSize(matrix_size_ * num_matrices_); //move to temp mem
+   lu_ptr_array_.SetSize(num_matrices_); //move to temp memory
+
+   // TODO: can this just be a Write?
+   double *lu_ptr_base = LUMatrixBatch_.ReadWrite();
+
+   const int matrix_size   = matrix_size_;
+   double **d_lu_ptr_array = lu_ptr_array_.Write();
+
+   mfem::forall(num_matrices_, [=] MFEM_HOST_DEVICE (int i)
+   {
+      d_lu_ptr_array[i] = lu_ptr_base + i * matrix_size * matrix_size;
+   });
+
+   switch (mode_)
+   {
+      case SolveMode::LU: ComputeLU(); break;
+
+      case SolveMode::INVERSE:
+         ComputeLU();
+         InvMatrixBatch_.SetSize(matrix_size_,
+                                 matrix_size_,
+                                 num_matrices_); //move to temporary memory
+         ComputeInverse(InvMatrixBatch_);
+         break;
+
+      default: mfem_error("Case not supported");
+   }
+   setup_ = true;
+}
+
+void LibBatchSolver::Mult(const mfem::Vector &b, mfem::Vector &x) const
+{
+   switch (mode_)
+   {
+      case SolveMode::LU: return SolveLU(b, x);
+
+      case SolveMode::INVERSE: return ApplyInverse(b, x);
+
+      default: mfem_error("Case not supported");
+   }
+}
+
+void LibBatchSolver::ReleaseMemory()
+{
+   MFEM_RELEASE_MEM(LUMatrixBatch_);
+   MFEM_RELEASE_MEM(InvMatrixBatch_);
+   MFEM_RELEASE_MEM(P_);
+   MFEM_RELEASE_MEM(lu_ptr_array_);
```

Comment:
I vote to remove the macro and just call `GetMemory().DeleteDevice(false)`, we don't know how others will want to use this. Please also change the name to `ReleaseDeviceMemory`.

Reply:
The name of the method? i.e. `DeleteDevice -> ReleaseDeviceMemory'?

Reply:
> The name of the method? i.e. `DeleteDevice -> ReleaseDeviceMemory'?

Yes please !